functorch/_src/eager_transforms.py

```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
         grad, _ = results
         return grad
     return wrapper
+
+
+def assert_wrapped_functional(
+        unwrapped_tensor_inputs: List[torch.Tensor],
+        wrapped_tensor_inputs: List[torch.Tensor]) -> None:
+    if len(unwrapped_tensor_inputs) != len(wrapped_tensor_inputs):
+        raise Exception()
+    for i in range(len(unwrapped_tensor_inputs)):
+        _assert_wrapped_functional(unwrapped_tensor_inputs[i], wrapped_tensor_inputs[i])
+
+
+def propagate_functional_input_mutations(
+        unwrapped_tensor_inputs: List[torch.Tensor],
+        wrapped_tensor_inputs: List[torch.Tensor]) -> None:
+    if len(unwrapped_tensor_inputs) != len(wrapped_tensor_inputs):
+        raise Exception()
+    for i in range(len(unwrapped_tensor_inputs)):
+        _propagate_functional_input_mutation(unwrapped_tensor_inputs[i], wrapped_tensor_inputs[i])
+
+
+def functionalize(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        try:
+            func_level = _func_increment_nesting()
+            func_args = [_wrap_functional_tensor(x, func_level) if isinstance(x, torch.Tensor) else x for x in args]
+
+            unwrapped_tensor_inputs = [t for t in args if isinstance(t, torch.Tensor)]
+            wrapped_tensor_inputs = [t for t in func_args if isinstance(t, torch.Tensor)]
+            assert_wrapped_functional(unwrapped_tensor_inputs, wrapped_tensor_inputs)
+
+            func_outputs = func(*func_args, **kwargs)
+            flattened_outputs, _ = tree_flatten(func_outputs)
+            for a in func_args:
+                if isinstance(a, torch.Tensor):
+                    # Call sync_() on the inputs, to ensure that they still get mutated inplace if the original
+                    # program mutated its inputs
+                    torch._sync(a)
+
+            propagate_functional_input_mutations(unwrapped_tensor_inputs, wrapped_tensor_inputs)
+            return [_unwrap_functional_tensor(x) if isinstance(x, torch.Tensor) else x for x in flattened_outputs]
+        finally:
+            _func_decrement_nesting()
+    return wrapped
```

Comment:
I added some suggestions for UX for this transform. tbh, when we started writing other transforms (e.g. vmap, grad), we just asserted that all inputs were tensors and that all outputs are tensors for simplicity so we didn't need to deal with tree_map. So feel free to ignore the comments above